File: tests/cpp/test_error_codes.cpp
/**
 * @file test_error_codes.cpp
 * @brief Unit tests for error codes and JSON error response building.
 */

#include "error_codes.h"
#include "zeromq_interface.h"

#include <gtest/gtest.h>
#include <nlohmann/json.hpp>

using json = nlohmann::json;
using namespace AudioEngine;

// ============================================================
// ErrorCode to String Tests
// ============================================================

TEST(ErrorCodes, ErrorCodeToString) {
    EXPECT_STREQ(errorCodeToString(ErrorCode::OK), "OK");
    EXPECT_STREQ(errorCodeToString(ErrorCode::DAC_RATE_NOT_SUPPORTED), "DAC_RATE_NOT_SUPPORTED");
    EXPECT_STREQ(errorCodeToString(ErrorCode::IPC_TIMEOUT), "IPC_TIMEOUT");
    EXPECT_STREQ(errorCodeToString(ErrorCode::GPU_MEMORY_ERROR), "GPU_MEMORY_ERROR");
    EXPECT_STREQ(errorCodeToString(ErrorCode::VALIDATION_PATH_TRAVERSAL),
                 "VALIDATION_PATH_TRAVERSAL");
}

TEST(ErrorCodes, UnknownErrorCodeReturnsUnknown) {
    auto unknownCode = static_cast<ErrorCode>(0xFFFF);
    EXPECT_STREQ(errorCodeToString(unknownCode), "UNKNOWN_ERROR");
}

// ============================================================
// Error Category Tests
// ============================================================

TEST(ErrorCodes, GetErrorCategory) {
    EXPECT_STREQ(getErrorCategory(ErrorCode::OK), "ok");
    EXPECT_STREQ(getErrorCategory(ErrorCode::AUDIO_INVALID_INPUT_RATE), "audio_processing");
    EXPECT_STREQ(getErrorCategory(ErrorCode::DAC_DEVICE_NOT_FOUND), "dac_alsa");
    EXPECT_STREQ(getErrorCategory(ErrorCode::IPC_CONNECTION_FAILED), "ipc_zeromq");
    EXPECT_STREQ(getErrorCategory(ErrorCode::GPU_INIT_FAILED), "gpu_cuda");
    EXPECT_STREQ(getErrorCategory(ErrorCode::VALIDATION_INVALID_CONFIG), "validation");
}

TEST(ErrorCodes, UnknownCategoryReturnsInternal) {
    auto unknownCode = static_cast<ErrorCode>(0xFFFF);
    EXPECT_STREQ(getErrorCategory(unknownCode), "internal");
}

// ============================================================
// Category Check Helper Tests
// ============================================================

TEST(ErrorCodes, CategoryCheckers) {
    EXPECT_TRUE(isAudioError(ErrorCode::AUDIO_INVALID_INPUT_RATE));
    EXPECT_TRUE(isAudioError(ErrorCode::AUDIO_XRUN_DETECTED));
    EXPECT_FALSE(isAudioError(ErrorCode::DAC_DEVICE_NOT_FOUND));

    EXPECT_TRUE(isDacError(ErrorCode::DAC_DEVICE_NOT_FOUND));
    EXPECT_TRUE(isDacError(ErrorCode::DAC_BUSY));
    EXPECT_FALSE(isDacError(ErrorCode::IPC_TIMEOUT));

    EXPECT_TRUE(isIpcError(ErrorCode::IPC_CONNECTION_FAILED));
    EXPECT_TRUE(isIpcError(ErrorCode::IPC_PROTOCOL_ERROR));
    EXPECT_FALSE(isIpcError(ErrorCode::GPU_INIT_FAILED));

    EXPECT_TRUE(isGpuError(ErrorCode::GPU_INIT_FAILED));
    EXPECT_TRUE(isGpuError(ErrorCode::GPU_CUFFT_ERROR));
    EXPECT_FALSE(isGpuError(ErrorCode::VALIDATION_INVALID_CONFIG));

    EXPECT_TRUE(isValidationError(ErrorCode::VALIDATION_INVALID_CONFIG));
    EXPECT_TRUE(isValidationError(ErrorCode::VALIDATION_INVALID_HEADPHONE));
    EXPECT_FALSE(isValidationError(ErrorCode::AUDIO_BUFFER_OVERFLOW));
}

// ============================================================
// HTTP Status Mapping Tests
// ============================================================

TEST(ErrorCodes, ToHttpStatus) {
    EXPECT_EQ(toHttpStatus(ErrorCode::OK), 200);

    // 400 Bad Request
    EXPECT_EQ(toHttpStatus(ErrorCode::AUDIO_INVALID_INPUT_RATE), 400);
    EXPECT_EQ(toHttpStatus(ErrorCode::IPC_INVALID_COMMAND), 400);
    EXPECT_EQ(toHttpStatus(ErrorCode::VALIDATION_INVALID_CONFIG), 400);

    // 404 Not Found
    EXPECT_EQ(toHttpStatus(ErrorCode::DAC_DEVICE_NOT_FOUND), 404);
    EXPECT_EQ(toHttpStatus(ErrorCode::AUDIO_FILTER_NOT_FOUND), 404);
    EXPECT_EQ(toHttpStatus(ErrorCode::VALIDATION_FILE_NOT_FOUND), 404);

    // 409 Conflict
    EXPECT_EQ(toHttpStatus(ErrorCode::DAC_BUSY), 409);
    EXPECT_EQ(toHttpStatus(ErrorCode::VALIDATION_PROFILE_EXISTS), 409);

    // 422 Unprocessable Entity
    EXPECT_EQ(toHttpStatus(ErrorCode::DAC_RATE_NOT_SUPPORTED), 422);
    EXPECT_EQ(toHttpStatus(ErrorCode::DAC_FORMAT_NOT_SUPPORTED), 422);

    // 500 Internal Server Error
    EXPECT_EQ(toHttpStatus(ErrorCode::GPU_INIT_FAILED), 500);
    EXPECT_EQ(toHttpStatus(ErrorCode::AUDIO_BUFFER_OVERFLOW), 500);

    // 503 Service Unavailable
    EXPECT_EQ(toHttpStatus(ErrorCode::IPC_CONNECTION_FAILED), 503);
    EXPECT_EQ(toHttpStatus(ErrorCode::IPC_DAEMON_NOT_RUNNING), 503);

    // 504 Gateway Timeout
    EXPECT_EQ(toHttpStatus(ErrorCode::IPC_TIMEOUT), 504);
}

TEST(ErrorCodes, UnknownErrorReturns500) {
    auto unknownCode = static_cast<ErrorCode>(0xFFFF);
    EXPECT_EQ(toHttpStatus(unknownCode), 500);
}

// ============================================================
// Error Code to Hex Tests
// ============================================================

TEST(ErrorCodes, ErrorCodeToHex) {
    EXPECT_EQ(errorCodeToHex(ErrorCode::OK), "0x0000");
    EXPECT_EQ(errorCodeToHex(ErrorCode::AUDIO_INVALID_INPUT_RATE), "0x1001");
    EXPECT_EQ(errorCodeToHex(ErrorCode::DAC_RATE_NOT_SUPPORTED), "0x2004");
    EXPECT_EQ(errorCodeToHex(ErrorCode::IPC_TIMEOUT), "0x3002");
    EXPECT_EQ(errorCodeToHex(ErrorCode::GPU_MEMORY_ERROR), "0x4003");
    EXPECT_EQ(errorCodeToHex(ErrorCode::VALIDATION_PATH_TRAVERSAL), "0x5003");
}

// ============================================================
// InnerError Tests
// ============================================================

TEST(ErrorCodes, InnerErrorConstruction) {
    InnerError err(ErrorCode::DAC_RATE_NOT_SUPPORTED, "Rate 1000000 not supported");

    EXPECT_EQ(err.cpp_code, "0x2004");
    EXPECT_EQ(err.cpp_message, "Rate 1000000 not supported");
    EXPECT_FALSE(err.alsa_errno.has_value());
    EXPECT_FALSE(err.alsa_func.has_value());
    EXPECT_FALSE(err.cuda_error.has_value());
}

TEST(ErrorCodes, InnerErrorWithAlsaInfo) {
    InnerError err;
    err.cpp_code = "0x2004";
    err.cpp_message = "Rate not supported";
    err.alsa_errno = -22;
    err.alsa_func = "snd_pcm_hw_params_set_rate_near";

    EXPECT_TRUE(err.alsa_errno.has_value());
    EXPECT_EQ(err.alsa_errno.value(), -22);
    EXPECT_TRUE(err.alsa_func.has_value());
    EXPECT_EQ(err.alsa_func.value(), "snd_pcm_hw_params_set_rate_near");
}

// ============================================================
// JSON Error Response Building Tests
// ============================================================

TEST(ErrorCodes, BuildErrorResponseBasic) {
    auto response =
        ZMQComm::JSON::buildErrorResponse(ErrorCode::DAC_RATE_NOT_SUPPORTED, "Rate not supported");

    auto j = json::parse(response);

    EXPECT_EQ(j["status"], "error");
    EXPECT_EQ(j["error_code"], "DAC_RATE_NOT_SUPPORTED");
    EXPECT_EQ(j["message"], "Rate not supported");
    EXPECT_TRUE(j.contains("inner_error"));
    EXPECT_EQ(j["inner_error"]["cpp_code"], "0x2004");
    // cpp_message should fallback to outer message when InnerError is default
    EXPECT_EQ(j["inner_error"]["cpp_message"], "Rate not supported");
}

TEST(ErrorCodes, BuildErrorResponseWithInnerError) {
    InnerError inner;
    inner.cpp_code = "0x2004";
    inner.cpp_message = "ALSA rate negotiation failed";
    inner.alsa_errno = -22;
    inner.alsa_func = "snd_pcm_hw_params_set_rate_near";

    auto response = ZMQComm::JSON::buildErrorResponse(ErrorCode::DAC_RATE_NOT_SUPPORTED,
                                                      "Rate 1000000 not supported", inner);

    auto j = json::parse(response);

    EXPECT_EQ(j["status"], "error");
    EXPECT_EQ(j["error_code"], "DAC_RATE_NOT_SUPPORTED");
    EXPECT_EQ(j["message"], "Rate 1000000 not supported");

    auto& innerJson = j["inner_error"];
    EXPECT_EQ(innerJson["cpp_code"], "0x2004");
    EXPECT_EQ(innerJson["cpp_message"], "ALSA rate negotiation failed");
    EXPECT_EQ(innerJson["alsa_errno"], -22);
    EXPECT_EQ(innerJson["alsa_func"], "snd_pcm_hw_params_set_rate_near");
    EXPECT_TRUE(innerJson["cuda_error"].is_null());
}

TEST(ErrorCodes, BuildErrorResponseWithCudaError) {
    InnerError inner;
    inner.cpp_code = "0x4003";
    inner.cpp_message = "Failed to allocate GPU memory";
    inner.cuda_error = "cudaErrorMemoryAllocation";

    auto response =
        ZMQComm::JSON::buildErrorResponse(ErrorCode::GPU_MEMORY_ERROR, "GPU memory error", inner);

    auto j = json::parse(response);

    EXPECT_EQ(j["error_code"], "GPU_MEMORY_ERROR");
    EXPECT_EQ(j["inner_error"]["cuda_error"], "cudaErrorMemoryAllocation");
    EXPECT_TRUE(j["inner_error"]["alsa_errno"].is_null());
}

// ============================================================
// JSON OK Response Building Tests
// ============================================================

TEST(ErrorCodes, BuildOkResponseBasic) {
    auto response = ZMQComm::JSON::buildOkResponse();

    auto j = json::parse(response);

    EXPECT_EQ(j["status"], "ok");
    EXPECT_FALSE(j.contains("message"));
    EXPECT_FALSE(j.contains("data"));
}

TEST(ErrorCodes, BuildOkResponseWithMessage) {
    auto response = ZMQComm::JSON::buildOkResponse("Operation completed");

    auto j = json::parse(response);

    EXPECT_EQ(j["status"], "ok");
    EXPECT_EQ(j["message"], "Operation completed");
}

TEST(ErrorCodes, BuildOkResponseWithData) {
    auto response = ZMQComm::JSON::buildOkResponse("Success", R"({"rate": 44100})");

    auto j = json::parse(response);

    EXPECT_EQ(j["status"], "ok");
    EXPECT_EQ(j["message"], "Success");
    EXPECT_EQ(j["data"]["rate"], 44100);
}

// ============================================================
// JSON Error Response Parsing Tests
// ============================================================

TEST(ErrorCodes, ParseErrorResponse) {
    std::string errorJson = R"({
        "status": "error",
        "error_code": "DAC_RATE_NOT_SUPPORTED",
        "message": "Rate not supported",
        "inner_error": {
            "cpp_code": "0x2004",
            "cpp_message": "ALSA error",
            "alsa_errno": -22
        }
    })";

    std::string errorCode, message, innerErrorJson;
    EXPECT_TRUE(ZMQComm::JSON::parseErrorResponse(errorJson, errorCode, message, innerErrorJson));

    EXPECT_EQ(errorCode, "DAC_RATE_NOT_SUPPORTED");
    EXPECT_EQ(message, "Rate not supported");

    auto inner = json::parse(innerErrorJson);
    EXPECT_EQ(inner["cpp_code"], "0x2004");
    EXPECT_EQ(inner["alsa_errno"], -22);
}

TEST(ErrorCodes, ParseErrorResponseFailsOnOkStatus) {
    std::string okJson = R"({"status": "ok", "message": "Success"})";

    std::string errorCode, message, innerErrorJson;
    EXPECT_FALSE(ZMQComm::JSON::parseErrorResponse(okJson, errorCode, message, innerErrorJson));
}

TEST(ErrorCodes, ParseErrorResponseFailsOnMissingErrorCode) {
    // Missing error_code field - should fail validation
    std::string brokenJson = R"({
        "status": "error",
        "message": "Some error"
    })";

    std::string errorCode, message, innerErrorJson;
    EXPECT_FALSE(ZMQComm::JSON::parseErrorResponse(brokenJson, errorCode, message, innerErrorJson));
}

TEST(ErrorCodes, ParseErrorResponseFailsOnMissingMessage) {
    // Missing message field - should fail validation
    std::string brokenJson = R"({
        "status": "error",
        "error_code": "IPC_TIMEOUT"
    })";

    std::string errorCode, message, innerErrorJson;
    EXPECT_FALSE(ZMQComm::JSON::parseErrorResponse(brokenJson, errorCode, message, innerErrorJson));
}

TEST(ErrorCodes, ParseErrorResponseFailsOnWrongTypeErrorCode) {
    // error_code is not a string - should fail validation
    std::string brokenJson = R"({
        "status": "error",
        "error_code": 12345,
        "message": "Some error"
    })";

    std::string errorCode, message, innerErrorJson;
    EXPECT_FALSE(ZMQComm::JSON::parseErrorResponse(brokenJson, errorCode, message, innerErrorJson));
}

TEST(ErrorCodes, ParseErrorResponseFailsOnWrongTypeMessage) {
    // message is not a string - should fail validation
    std::string brokenJson = R"({
        "status": "error",
        "error_code": "IPC_TIMEOUT",
        "message": 12345
    })";

    std::string errorCode, message, innerErrorJson;
    EXPECT_FALSE(ZMQComm::JSON::parseErrorResponse(brokenJson, errorCode, message, innerErrorJson));
}

// ============================================================
// All 30 Error Codes Coverage Test
// ============================================================

TEST(ErrorCodes, AllErrorCodesHaveStringMapping) {
    // Audio Processing (6 codes)
    EXPECT_STRNE(errorCodeToString(ErrorCode::AUDIO_INVALID_INPUT_RATE), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::AUDIO_INVALID_OUTPUT_RATE), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::AUDIO_UNSUPPORTED_FORMAT), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::AUDIO_FILTER_NOT_FOUND), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::AUDIO_BUFFER_OVERFLOW), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::AUDIO_XRUN_DETECTED), "UNKNOWN_ERROR");

    // DAC/ALSA (6 codes)
    EXPECT_STRNE(errorCodeToString(ErrorCode::DAC_DEVICE_NOT_FOUND), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::DAC_OPEN_FAILED), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::DAC_CAPABILITY_SCAN_FAILED), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::DAC_RATE_NOT_SUPPORTED), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::DAC_FORMAT_NOT_SUPPORTED), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::DAC_BUSY), "UNKNOWN_ERROR");

    // IPC/ZeroMQ (6 codes)
    EXPECT_STRNE(errorCodeToString(ErrorCode::IPC_CONNECTION_FAILED), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::IPC_TIMEOUT), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::IPC_INVALID_COMMAND), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::IPC_INVALID_PARAMS), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::IPC_DAEMON_NOT_RUNNING), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::IPC_PROTOCOL_ERROR), "UNKNOWN_ERROR");

    // GPU/CUDA (6 codes)
    EXPECT_STRNE(errorCodeToString(ErrorCode::GPU_INIT_FAILED), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::GPU_DEVICE_NOT_FOUND), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::GPU_MEMORY_ERROR), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::GPU_KERNEL_LAUNCH_FAILED), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::GPU_FILTER_LOAD_FAILED), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::GPU_CUFFT_ERROR), "UNKNOWN_ERROR");

    // Validation (6 codes)
    EXPECT_STRNE(errorCodeToString(ErrorCode::VALIDATION_INVALID_CONFIG), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::VALIDATION_INVALID_PROFILE), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::VALIDATION_PATH_TRAVERSAL), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::VALIDATION_FILE_NOT_FOUND), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::VALIDATION_PROFILE_EXISTS), "UNKNOWN_ERROR");
    EXPECT_STRNE(errorCodeToString(ErrorCode::VALIDATION_INVALID_HEADPHONE), "UNKNOWN_ERROR");
}

TEST(ErrorCodes, AllErrorCodesHaveHttpStatusMapping) {
    // All 30 codes should have HTTP status (not default 500 unless intentional)
    // Just verify they all return a valid HTTP status code
    auto validateStatus = [](ErrorCode code) {
        int status = toHttpStatus(code);
        return status >= 200 && status < 600;
    };

    EXPECT_TRUE(validateStatus(ErrorCode::AUDIO_INVALID_INPUT_RATE));
    EXPECT_TRUE(validateStatus(ErrorCode::DAC_RATE_NOT_SUPPORTED));
    EXPECT_TRUE(validateStatus(ErrorCode::IPC_TIMEOUT));
    EXPECT_TRUE(validateStatus(ErrorCode::GPU_MEMORY_ERROR));
    EXPECT_TRUE(validateStatus(ErrorCode::VALIDATION_PATH_TRAVERSAL));
}

// ============================================================
// INTERNAL_UNKNOWN Tests (Issue #44 additions)
// ============================================================

TEST(ErrorCodes, InternalUnknownExists) {
    EXPECT_STREQ(errorCodeToString(ErrorCode::INTERNAL_UNKNOWN), "INTERNAL_UNKNOWN");
    EXPECT_EQ(errorCodeToHex(ErrorCode::INTERNAL_UNKNOWN), "0xf001");
    EXPECT_EQ(toHttpStatus(ErrorCode::INTERNAL_UNKNOWN), 500);
}

TEST(ErrorCodes, IsInternalErrorHelper) {
    EXPECT_TRUE(isInternalError(ErrorCode::INTERNAL_UNKNOWN));
    EXPECT_FALSE(isInternalError(ErrorCode::DAC_DEVICE_NOT_FOUND));
    EXPECT_FALSE(isInternalError(ErrorCode::GPU_INIT_FAILED));
    EXPECT_FALSE(isInternalError(ErrorCode::OK));
}

// ============================================================
// isRetryable Tests
// ============================================================

TEST(ErrorCodes, IsRetryableReturnsTrue) {
    EXPECT_TRUE(isRetryable(ErrorCode::IPC_DAEMON_NOT_RUNNING));
    EXPECT_TRUE(isRetryable(ErrorCode::IPC_TIMEOUT));
    EXPECT_TRUE(isRetryable(ErrorCode::IPC_CONNECTION_FAILED));
}

TEST(ErrorCodes, IsRetryableReturnsFalse) {
    EXPECT_FALSE(isRetryable(ErrorCode::OK));
    EXPECT_FALSE(isRetryable(ErrorCode::DAC_DEVICE_NOT_FOUND));
    EXPECT_FALSE(isRetryable(ErrorCode::GPU_MEMORY_ERROR));
    EXPECT_FALSE(isRetryable(ErrorCode::VALIDATION_INVALID_CONFIG));
    EXPECT_FALSE(isRetryable(ErrorCode::IPC_INVALID_COMMAND));
    EXPECT_FALSE(isRetryable(ErrorCode::INTERNAL_UNKNOWN));
}

// ============================================================
// stringToErrorCode Tests
// ============================================================

TEST(ErrorCodes, StringToErrorCodeValid) {
    EXPECT_EQ(stringToErrorCode("OK"), ErrorCode::OK);
    EXPECT_EQ(stringToErrorCode("AUDIO_INVALID_INPUT_RATE"), ErrorCode::AUDIO_INVALID_INPUT_RATE);
    EXPECT_EQ(stringToErrorCode("DAC_DEVICE_NOT_FOUND"), ErrorCode::DAC_DEVICE_NOT_FOUND);
    EXPECT_EQ(stringToErrorCode("IPC_TIMEOUT"), ErrorCode::IPC_TIMEOUT);
    EXPECT_EQ(stringToErrorCode("GPU_MEMORY_ERROR"), ErrorCode::GPU_MEMORY_ERROR);
    EXPECT_EQ(stringToErrorCode("VALIDATION_PATH_TRAVERSAL"), ErrorCode::VALIDATION_PATH_TRAVERSAL);
    EXPECT_EQ(stringToErrorCode("INTERNAL_UNKNOWN"), ErrorCode::INTERNAL_UNKNOWN);
}

TEST(ErrorCodes, StringToErrorCodeInvalid) {
    EXPECT_EQ(stringToErrorCode("INVALID_CODE"), ErrorCode::INTERNAL_UNKNOWN);
    EXPECT_EQ(stringToErrorCode(""), ErrorCode::INTERNAL_UNKNOWN);
    EXPECT_EQ(stringToErrorCode("random_string"), ErrorCode::INTERNAL_UNKNOWN);
}

TEST(ErrorCodes, RoundTripStringConversion) {
    // Test that errorCodeToString and stringToErrorCode are inverses
    auto testRoundTrip = [](ErrorCode code) {
        const char* str = errorCodeToString(code);
        ErrorCode result = stringToErrorCode(str);
        EXPECT_EQ(code, result) << "Failed for: " << str;
    };

    testRoundTrip(ErrorCode::OK);
    testRoundTrip(ErrorCode::AUDIO_XRUN_DETECTED);
    testRoundTrip(ErrorCode::DAC_BUSY);
    testRoundTrip(ErrorCode::IPC_PROTOCOL_ERROR);
    testRoundTrip(ErrorCode::GPU_CUFFT_ERROR);
    testRoundTrip(ErrorCode::VALIDATION_INVALID_HEADPHONE);
    testRoundTrip(ErrorCode::INTERNAL_UNKNOWN);
}

File: tests/python/test_error_codes.py
"""Tests for error code system.

Tests the error code enums, mappings, and utility functions
defined in web/error_codes.py.
"""

import pytest

from web.error_codes import (
    ERROR_MAPPINGS,
    ErrorCategory,
    ErrorCode,
    ErrorMapping,
    get_category_from_code,
    get_error_mapping,
)


class TestErrorCodeEnum:
    """Tests for ErrorCode enum."""

    def test_all_codes_defined(self):
        """Verify all error codes are defined (30 base + 4 crossfeed + 1 output)."""
        assert len(ErrorCode) == 35

    def test_no_duplicate_values(self):
        """Verify no duplicate enum values."""
        values = [code.value for code in ErrorCode]
        assert len(values) == len(set(values))

    def test_audio_processing_codes(self):
        """Verify Audio Processing category codes (6 codes)."""
        audio_codes = [c for c in ErrorCode if c.value.startswith("AUDIO_")]
        assert len(audio_codes) == 6
        assert ErrorCode.AUDIO_INVALID_INPUT_RATE in audio_codes
        assert ErrorCode.AUDIO_XRUN_DETECTED in audio_codes

    def test_dac_codes(self):
        """Verify DAC/ALSA category codes (6 codes)."""
        dac_codes = [c for c in ErrorCode if c.value.startswith("DAC_")]
        assert len(dac_codes) == 6
        assert ErrorCode.DAC_DEVICE_NOT_FOUND in dac_codes
        assert ErrorCode.DAC_BUSY in dac_codes

    def test_ipc_codes(self):
        """Verify IPC/ZeroMQ category codes (6 codes)."""
        ipc_codes = [c for c in ErrorCode if c.value.startswith("IPC_")]
        assert len(ipc_codes) == 6
        assert ErrorCode.IPC_TIMEOUT in ipc_codes
        assert ErrorCode.IPC_DAEMON_NOT_RUNNING in ipc_codes

    def test_gpu_codes(self):
        """Verify GPU/CUDA category codes (6 codes)."""
        gpu_codes = [c for c in ErrorCode if c.value.startswith("GPU_")]
        assert len(gpu_codes) == 6
        assert ErrorCode.GPU_MEMORY_ERROR in gpu_codes
        assert ErrorCode.GPU_CUFFT_ERROR in gpu_codes

    def test_validation_codes(self):
        """Verify Validation category codes (6 codes)."""
        validation_codes = [c for c in ErrorCode if c.value.startswith("VALIDATION_")]
        assert len(validation_codes) == 6
        assert ErrorCode.VALIDATION_PATH_TRAVERSAL in validation_codes
        assert ErrorCode.VALIDATION_INVALID_HEADPHONE in validation_codes

    def test_crossfeed_codes(self):
        """Verify Crossfeed/HRTF category codes (4 codes)."""
        crossfeed_codes = [c for c in ErrorCode if c.value.startswith("CROSSFEED_")]
        assert len(crossfeed_codes) == 4
        assert ErrorCode.CROSSFEED_NOT_INITIALIZED in crossfeed_codes
        assert ErrorCode.CROSSFEED_INVALID_RATE_FAMILY in crossfeed_codes
        assert ErrorCode.CROSSFEED_NOT_IMPLEMENTED in crossfeed_codes
        assert ErrorCode.CROSSFEED_INVALID_FILTER_SIZE in crossfeed_codes


class TestErrorCategory:
    """Tests for ErrorCategory enum."""

    def test_all_categories_defined(self):
        """Verify all categories are defined (6 base + 1 crossfeed)."""
        assert len(ErrorCategory) == 7

    def test_category_values(self):
        """Verify category string values."""
        assert ErrorCategory.AUDIO_PROCESSING.value == "audio_processing"
        assert ErrorCategory.DAC_ALSA.value == "dac_alsa"
        assert ErrorCategory.IPC_ZEROMQ.value == "ipc_zeromq"
        assert ErrorCategory.GPU_CUDA.value == "gpu_cuda"
        assert ErrorCategory.VALIDATION.value == "validation"
        assert ErrorCategory.CROSSFEED.value == "crossfeed"
        assert ErrorCategory.INTERNAL.value == "internal"


class TestErrorMappings:
    """Tests for ERROR_MAPPINGS dictionary."""

    def test_all_codes_have_mappings(self):
        """Verify all error codes have mappings (30 base + 4 crossfeed + 1 output)."""
        assert len(ERROR_MAPPINGS) == 35
        for code in ErrorCode:
            assert code in ERROR_MAPPINGS, f"Missing mapping for {code}"

    def test_mapping_structure(self):
        """Verify mapping structure is correct."""
        for code, mapping in ERROR_MAPPINGS.items():
            assert isinstance(mapping, ErrorMapping)
            assert isinstance(mapping.http_status, int)
            assert isinstance(mapping.category, ErrorCategory)
            assert isinstance(mapping.title, str)

    def test_http_status_ranges(self):
        """Verify HTTP status codes are in valid ranges."""
        valid_statuses = {400, 404, 409, 422, 500, 501, 503, 504}
        for code, mapping in ERROR_MAPPINGS.items():
            assert (
                mapping.http_status in valid_statuses
            ), f"{code} has invalid HTTP status {mapping.http_status}"

    @pytest.mark.parametrize(
        "error_code,expected_status",
        [
            (ErrorCode.AUDIO_INVALID_INPUT_RATE, 400),
            (ErrorCode.AUDIO_FILTER_NOT_FOUND, 404),
            (ErrorCode.DAC_DEVICE_NOT_FOUND, 404),
            (ErrorCode.DAC_RATE_NOT_SUPPORTED, 422),
            (ErrorCode.DAC_BUSY, 409),
            (ErrorCode.IPC_TIMEOUT, 504),
            (ErrorCode.IPC_DAEMON_NOT_RUNNING, 503),
            (ErrorCode.GPU_MEMORY_ERROR, 500),
            (ErrorCode.VALIDATION_PATH_TRAVERSAL, 400),
            (ErrorCode.VALIDATION_FILE_NOT_FOUND, 404),
            (ErrorCode.VALIDATION_PROFILE_EXISTS, 409),
        ],
    )
    def test_specific_status_mappings(self, error_code, expected_status):
        """Test specific error code to HTTP status mappings."""
        assert ERROR_MAPPINGS[error_code].http_status == expected_status

    @pytest.mark.parametrize(
        "error_code,expected_category",
        [
            (ErrorCode.AUDIO_BUFFER_OVERFLOW, ErrorCategory.AUDIO_PROCESSING),
            (ErrorCode.DAC_OPEN_FAILED, ErrorCategory.DAC_ALSA),
            (ErrorCode.IPC_CONNECTION_FAILED, ErrorCategory.IPC_ZEROMQ),
            (ErrorCode.GPU_INIT_FAILED, ErrorCategory.GPU_CUDA),
            (ErrorCode.VALIDATION_INVALID_CONFIG, ErrorCategory.VALIDATION),
        ],
    )
    def test_category_mappings(self, error_code, expected_category):
        """Test error code to category mappings."""
        assert ERROR_MAPPINGS[error_code].category == expected_category


class TestGetErrorMapping:
    """Tests for get_error_mapping() function."""

    def test_known_error_code(self):
        """Test lookup of known error code."""
        mapping = get_error_mapping("DAC_RATE_NOT_SUPPORTED")
        assert mapping.http_status == 422
        assert mapping.category == ErrorCategory.DAC_ALSA
        assert mapping.title == "DAC Rate Not Supported"

    def test_unknown_error_code(self):
        """Test fallback for unknown error code."""
        mapping = get_error_mapping("UNKNOWN_ERROR_CODE")
        assert mapping.http_status == 500
        assert mapping.category == ErrorCategory.INTERNAL
        assert mapping.title == "Internal Error"

    def test_empty_string(self):
        """Test fallback for empty string."""
        mapping = get_error_mapping("")
        assert mapping.http_status == 500
        assert mapping.category == ErrorCategory.INTERNAL

    def test_case_sensitivity(self):
        """Test that error codes are case-sensitive."""
        # Lowercase should not match
        mapping = get_error_mapping("dac_rate_not_supported")
        assert mapping.http_status == 500  # Falls back to default


class TestGetCategoryFromCode:
    """Tests for get_category_from_code() function."""

    @pytest.mark.parametrize(
        "error_code,expected_category",
        [
            ("AUDIO_SOMETHING", ErrorCategory.AUDIO_PROCESSING),
            ("DAC_WHATEVER", ErrorCategory.DAC_ALSA),
            ("IPC_ANYTHING", ErrorCategory.IPC_ZEROMQ),
            ("GPU_ERROR", ErrorCategory.GPU_CUDA),
            ("VALIDATION_ISSUE", ErrorCategory.VALIDATION),
        ],
    )
    def test_prefix_detection(self, error_code, expected_category):
        """Test category detection from error code prefix."""
        assert get_category_from_code(error_code) == expected_category

    def test_unknown_prefix(self):
        """Test fallback for unknown prefix."""
        assert get_category_from_code("UNKNOWN_ERROR") == ErrorCategory.INTERNAL
        assert get_category_from_code("") == ErrorCategory.INTERNAL


class TestErrorMappingDataclass:
    """Tests for ErrorMapping dataclass."""

    def test_frozen(self):
        """Verify ErrorMapping is immutable (frozen)."""
        mapping = ErrorMapping(
            http_status=400,
            category=ErrorCategory.VALIDATION,
            title="Test Error",
        )
        with pytest.raises(Exception):  # FrozenInstanceError
            mapping.http_status = 500

    def test_equality(self):
        """Test ErrorMapping equality."""
        mapping1 = ErrorMapping(400, ErrorCategory.VALIDATION, "Test")
        mapping2 = ErrorMapping(400, ErrorCategory.VALIDATION, "Test")
        assert mapping1 == mapping2

    def test_hashable(self):
        """Test ErrorMapping is hashable (can be used in sets/dicts)."""
        mapping = ErrorMapping(400, ErrorCategory.VALIDATION, "Test")
        # Should not raise
        hash(mapping)
        {mapping}  # Can be added to set
